File: Python/HeapSort.py
# Python Implementation of Heap Sort 
# Author : Aditya Mathur
# Date : 4 Oct 2020

def heapify(arr, n, i): 
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2
  
    # See if left child of root exists and is 
    # greater than root 
    if left< n and arr[i] < arr[left]: 
        largest = left
  
    # See if right child of root exists and is 
    # greater than root 
    if right < n and arr[largest] < arr[right]: 
        largest = right
  
    # Change root, if needed 
    if largest != i: 
        arr[i],arr[largest] = arr[largest],arr[i]
  
        # Heapify the root. 
        heapify(arr, n, largest) 

def heapSort(arr):
    n = len(arr)

    # Max Heap
    # Since last parent will be at ((n//2)-1) we can start at that location
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    # One by one extract elements 
    for i in range(n-1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]
        heapify(arr, i, 0) 


# Testing Code
arr = [ 56, 77, 2, 15, 30, 9]
heapSort(arr) 
print(" ".join(map(str, arr)))

File: C++/heap_sort.cpp
#include <iostream>
#include<vector>
using namespace std;
void max_heapify (vector<int>& Arr, int i,int N)
{
    int left = 2*i ;                  
    int right = 2*i +1 ;
    int largest=-1;
    if(left<= N and Arr[left] > Arr[i] )
        largest = left;
    else
        largest = i;
    if(right <= N and Arr[right] > Arr[largest] )
        largest = right;
    if(largest != i )
    {
        swap (Arr[i] , Arr[largest]);
        max_heapify (Arr, largest,N);
    }
}
void build_maxheap (vector<int> Arr)
{
	int N=Arr.size();
    for(int i = N/2 ; i >= 1 ; i-- )
    {
        max_heapify (Arr, i,N) ;
    }
}


void heap_sort(vector<int> &Arr)
{
    int heap_size = Arr.size();
    build_maxheap(Arr);
    for(int i = heap_size; i>=2 ; i-- )
    {
        swap(Arr[ 1 ], Arr[ i ]);
        heap_size = heap_size-1;
        max_heapify(Arr, 1, heap_size);
    }
}
int main() {
	int n=0;
	cin>>n;
	vector<int>a(n);
	for(int i=0;i<n;++i) cin>>a[i];
	heap_sort(a);
	for(auto x: a) cout<<x<<" ";
	
	
}
